utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,601 @@
+  "comp"  -> Just Comp
+  "tesc"  -> Just Tesc
+  "tese"  -> Just Tese
+  "geom"  -> Just Geom
+  "rgen"  -> Just Rgen
+  "rint"  -> Just Rint
+  "rahit" -> Just Rahit
+  "rchit" -> Just Rchit
+  "rmiss" -> Just Rmiss
+  "rcall" -> Just Rcall
+  "task"  -> Just Task
+  "mesh"  -> Just Mesh
+  _       -> Nothing
+
+shaderStageToText :: ShaderStage -> Text
+shaderStageToText = \case
+  Vert  -> "vert"
+  Frag  -> "frag"
+  Comp  -> "comp"
+  Tesc  -> "tesc"
+  Tese  -> "tese"
+  Geom  -> "geom"
+  Rgen  -> "rgen"
+  Rint  -> "rint"
+  Rahit -> "rahit"
+  Rchit -> "rchit"
+  Rmiss -> "rmiss"
+  Rcall -> "rcall"
+  Task  -> "task"
+  Mesh  -> "mesh"
+
+instance FromJSON ShaderStage where
+  parseJSON = withTextMaybe "mode" shaderStageFromText
+
+instance ToJSON ShaderStage where
+  toJSON     = String . shaderStageToText
+  toEncoding = toEncoding . shaderStageToText
+
+shaderStageFlagBits :: ShaderStage -> ShaderStageFlagBits
+shaderStageFlagBits = \case
+  Vert  -> SHADER_STAGE_VERTEX_BIT
+  Frag  -> SHADER_STAGE_FRAGMENT_BIT
+  Comp  -> SHADER_STAGE_COMPUTE_BIT
+  Tesc  -> SHADER_STAGE_TESSELLATION_CONTROL_BIT
+  Tese  -> SHADER_STAGE_TESSELLATION_EVALUATION_BIT
+  Geom  -> SHADER_STAGE_GEOMETRY_BIT
+  Rgen  -> SHADER_STAGE_RAYGEN_BIT_KHR
+  Rint  -> SHADER_STAGE_INTERSECTION_BIT_KHR
+  Rahit -> SHADER_STAGE_ANY_HIT_BIT_KHR
+  Rchit -> SHADER_STAGE_CLOSEST_HIT_BIT_KHR
+  Rmiss -> SHADER_STAGE_MISS_BIT_KHR
+  Rcall -> SHADER_STAGE_CALLABLE_BIT_KHR
+  Task  -> SHADER_STAGE_TASK_BIT_NV
+  Mesh  -> SHADER_STAGE_MESH_BIT_NV
+
+data Shader = Shader
+  { stage :: ShaderStage
+  , code  :: B.ByteString
+  }
+  deriving Show
+
+data ShaderInfo = ShaderInfo
+  { shaderModuleCreateInfo :: ShaderModuleCreateInfo '[]
+  , pipelineShaderStageCreateInfos
+      :: ShaderModule -> Vector (SomeStruct PipelineShaderStageCreateInfo)
+  }
+
+makeShaderInfo :: (Shader, Reflection) -> ShaderInfo
+makeShaderInfo (Shader {..}, Reflection {..}) = ShaderInfo { .. }
+ where
+  shaderModuleCreateInfo = makeShaderModuleCreateInfo code
+  pipelineShaderStageCreateInfos =
+    (SomeStruct <$>) . makePipelineShaderStageCreateInfos stage entryPoints
+
+makeDescriptorInfo
+  :: Vector (Shader, Reflection) -> Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorInfo = makeDescriptorSetLayoutCreateInfos . join . V.map
+  (   makeDescriptorSetLayoutBindings
+  <$> makeShaderStage
+  <*> makeUbos
+  <*> makeTextures
+  )
+ where
+  makeShaderStage :: (Shader, Reflection) -> ShaderStage
+  makeShaderStage = (stage :: Shader -> ShaderStage) . fst
+  makeUbos :: (Shader, Reflection) -> Vector Ubo
+  makeUbos = fromMaybe [] . ubos . snd
+  makeTextures :: (Shader, Reflection) -> Vector Texture
+  makeTextures = fromMaybe [] . textures . snd
+
+makeInputInfo
+  :: Vector (Shader, Reflection)
+  -> Maybe (SomeStruct PipelineVertexInputStateCreateInfo)
+makeInputInfo =
+  (SomeStruct <$>)
+    . makePipelineVertexInputStateCreateInfo
+    . join
+    . V.fromList
+    . catMaybes
```

Comment:
Vector has `mapMaybe`. This will spare you from using toList/fromList.

Reply:
fixed

Reply:
Hm... this brings us back to `catMaybe`, but this time, it is from `vector`.

Reply:
fixed. now its `V.mapMaybe (inputs . snd)`